ci: Setup gitflow process

mydea opened the pull request:
This PR (WIP) sets up some Gitflow processes.

Some requirements that need to be done in order for this to work/make sense:

* Extend all branch protection settings we have for `master` to `develop` (so have them for both branches)
* Make `develop` the default branch in Github
* Either make sure that approval of the  "github-actions" bot user is OK to merge a PR, or we need to provide a github token of a user in whose name the approvals will be in

## How it should work (overview)

* We develop on the new `develop` branch. Any PRs are merged into develop
* When we want to do a release, we merge develop into master and trigger a release from there.
* While a release is pending, PRs can still be merged into develop (they will just not go into the ongoing release anymore)

![gitflow-sentry-javascript (1)](https://user-images.githubusercontent.com/2411343/213697023-276d878d-f06e-4f89-bf82-6e7b2da07c1f.png)

## How to do a release going forward?

* Create a meta changelog PR to develop (as before)
* When this is merged, we'll automatically create a PR from develop->master, auto-approve it, and automerge it when CI passed
* Trigger a release from master branch (as before)
* When this is merged, we'll automatically create a PR from master->develop, auto-approve it, and automerge it when CI passed
* Now everything should be in sync again!

## Potential streamlining/improvement?

One way in which this _could_ be improved, is to auto-trigger the release whenever we merge into master. 
In order to do this, we'll need to somehow know at this point:

* the version to release
* force if there are release blockers (optional)

We may be able to parse this somehow, but I'd say this is an area for future improvement, if we decide so.

## What's missing?

We'll probably have to tweak caching etc. settings a bit to streamline this - e.g. what needs to run when merging develop-> master, what needs to run when merging master->develop, etc.

Lms24 reviewed (commented):
Seems like a good workflow to me!

Also, great diagram! Once we all agree on this, let's make sure to document this flow in our docs directory and maybe link to it from `CONTRIBUTING` or similar?

mydea commented:
Added docs here: https://github.com/getsentry/sentry-javascript/blob/fn/gitflow/CONTRIBUTING.md

Note: I took this opportunity and extracted the publishing stuff out as well.

Lms24 commented on docs/publishing-a-release.md:
```diff
@@ -0,0 +1,24 @@
+**If you want to release a new SDK for the first time, be sure to follow the [New SDK Release Checklist](./new-sdk-release-checklist.md)**
+
+1. Determine what version will be released (we use [semver](https://semver.org)).
+2. Update [`CHANGELOG.md`](https://github.com/getsentry/sentry-javascript/edit/master/CHANGELOG.md) to add an entry for the next release number and a list of changes since the last release. (See details below.)
```
the way I understand it, the changelog PR merge will trigger the merge from develop into master, right? Let's maybe add a comment to point this out. I think this might be valuable information for anyone coming into the repo.

mydea replied:
I updated it to point this out specifically, is that what you had in mind?

Lms24 replied:
Yes, I was thinking of exactly something like this. Thanks!

Lms24 reviewed (approved):
LGTM. I like the automatic approach master/develop syncing approach and I think going forward, moving to gitflow will solve our release blocking/multiple teams working on the repo problem. Thanks for updating documentation, I think it makes a lot of sense now.

mydea commented:
FYI I updated this slightly:

1. Added the gitflow image to the docs
2. Made it so we can manually trigger a sync between develop/master via GHA workflow_dispatch
